src/GafferScene/ExistenceQuery.cpp

```diff
@@ -0,0 +1,252 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2021, Cinesite VFX Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferScene/ExistenceQuery.h"
+
+#include <cassert>
+
+namespace GafferScene
+{
+
+size_t ExistenceQuery::g_firstPlugIndex = 0;
+
+GAFFER_NODE_DEFINE_TYPE( ExistenceQuery );
+
+ExistenceQuery::ExistenceQuery( std::string const& name )
+: Gaffer::ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new ScenePlug( "scene" ) );
+	addChild( new Gaffer::StringPlug( "location" ) );
+	addChild( new Gaffer::BoolPlug( "exists", Gaffer::Plug::Out, false ) );
+	addChild( new Gaffer::StringPlug( "closestAncestor", Gaffer::Plug::Out ) );
+}
+
+ExistenceQuery::~ExistenceQuery()
+{}
+
+ScenePlug* ExistenceQuery::scenePlug()
+{
+	return const_cast< ScenePlug* >(
+		static_cast< ExistenceQuery const* >( this )->scenePlug() );
+}
+
+ScenePlug const* ExistenceQuery::scenePlug() const
+{
+	return getChild< ScenePlug >( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug* ExistenceQuery::locationPlug()
+{
+	return const_cast< Gaffer::StringPlug* >(
```

Comment:
Hmm, interesting to use this cast to get rid of the duplicate definition of the plug index for const/non-const that we usually have in our node boilerplate.  I don't mind it, though I wonder if we should be consistent for all our nodes.

I do find that the amount of boilerplate is genuinely pretty distracting when looking at a node ... I would be pretty up for just writing a horrible macro that let us do a one liner that would declare both, something like:
`DECLARE_GAFFER_PLUG_ACCESSORS( ExistenceQuery::locationPlug, Gaffer::StringPlug, 1 )`
But I guess John has been OK with the current verbose boilerplate.

Reply:
I'm not especially keen on using a macro in the headers, because as a reader of the headers I find it easier to look at plain old code. What I find most painful in the `.cpp` files is managing the indices, particularly when it comes to adding new plugs in the middle at a later date. A little googling suggests it's _maybe_ it's possible to do some horrendous macro with an internal counter to do it in one shot like this :

```
GAFFER_NODE_DEFINE_PLUG_ACCESSORS(
    ExistenceQuery::scenePlug, GafferScene::ScenePlug,
    ExistenceQuery::locationPlug, Gaffer::StringPlug,
    ...
)
```

Probably way off in the future, but metaclasses sound like they might give us a far more elegant way of dealing with this : https://herbsutter.files.wordpress.com/2017/07/p0707r1.pdf.

In any case, not something to worry about on this PR.